### PR TITLE
Add legalization for missing ops from LHLO to LinAlg.

### PR DESCRIPTION
**Pull request by ehsan-nt**

### CODE REVIEW

File: tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h [hunk 1]
```diff
@@ -59,69 +64,71 @@ template <typename LHLO_BinaryOp>
...
 
 template <typename LhloOp>
-Operation* MapLhloOpToStdScalarOp(LhloOp lhlo_op, ArrayRef<Type> result_types,
-                                  ArrayRef<Value*> block_args, OpBuilder b) {
+Value* MapLhloOpToStdScalarOp(LhloOp lhlo_op, ArrayRef<Type> result_types,
+                              ArrayRef<Value*> block_args, OpBuilder b) {
   Type element_type = block_args.front()->getType();
+  Operation* op;
   if (element_type.isa<IntegerType>()) {
-    return b.template create<ScalarIOp<LhloOp>>(lhlo_op.getLoc(), result_types,
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h]: Why not `return b.template create<ScalarIOp<LhloOp>>(lhlo_op.getLoc(), result_types, block_args, mlir::None).getResult()`. I would find that easier to read compared to the extra conditional at the end. You should be able to use getResult() here, as the created operations have a single result.

File: tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h [hunk 2]
```diff
@@ -59,69 +64,71 @@ template <typename LHLO_BinaryOp>
...
 
 template <>
-inline Operation* MapLhloOpToStdScalarOp<xla_lhlo::MaxOp>(
+inline Value* MapLhloOpToStdScalarOp<xla_lhlo::MaxOp>(
     xla_lhlo::MaxOp lhlo_op, ArrayRef<Type> result_types,
     ArrayRef<Value*> block_args, OpBuilder b) {
   const auto& lhs = block_args[0];
   const auto& rhs = block_args[1];
   Type element_type = lhs->getType();
+  Operation* op;
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h]: Same comment here. Why not return with getResult() directly?

File: tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h [hunk 3]
```diff
@@ -59,69 +64,71 @@ template <typename LHLO_BinaryOp>
...
+inline Value* MapLhloOpToStdScalarOp<xla_lhlo::AndOp>(
     xla_lhlo::AndOp lhlo_op, ArrayRef<Type> result_types,
     ArrayRef<Value*> block_args, OpBuilder b) {
   Type element_type = block_args.front()->getType();
-  return element_type.isa<IntegerType>()
-             ? b.create<::mlir::AndOp>(lhlo_op.getLoc(), result_types,
-                                       block_args, mlir::None)
-             : nullptr;
+  if (element_type.isa<IntegerType>()) {
+    Operation* op = b.create<::mlir::AndOp>(lhlo_op.getLoc(), result_types,
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h]: If you avoid the conversion to Operation* here, you can use getResult().

File: tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h [hunk 4]
```diff
@@ -148,44 +155,173 @@ inline Optional<CmpIPredicate> getIntCmpPredicate(
...
 }
 
 template <>
-inline Operation* MapLhloOpToStdScalarOp<xla_lhlo::ExpOp>(
+inline Value* MapLhloOpToStdScalarOp<xla_lhlo::CeilOp>(
+    xla_lhlo::CeilOp lhlo_op, ArrayRef<Type> result_types,
+    ArrayRef<Value*> block_args, OpBuilder b) {
+  Type element_type = block_args.front()->getType();
+  if (element_type.isa<FloatType>()) {
+    Operation* op = b.create<::mlir::CeilFOp>(lhlo_op.getLoc(), result_types,
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h]: Generally, avoid converting to `Operation *`. Either use auto here or the specific type (`CeilFOp`) to get access to operation specific accessors like getResult(). This comment applies to all templates here.

File: tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h [hunk 5]
```diff
@@ -148,44 +155,173 @@ inline Optional<CmpIPredicate> getIntCmpPredicate(
...
+    FloatType src = sourceType.cast<FloatType>();
+    FloatType res = targetType.cast<FloatType>();
+    if (src.getWidth() > res.getWidth()) {
+      op = b.create<mlir::FPTruncOp>(lhlo_op.getLoc(), result_types, block_args,
+                                     mlir::None);
+    } else if (src.getWidth() < res.getWidth()) {
+      op = b.create<mlir::FPExtOp>(lhlo_op.getLoc(), result_types, block_args,
+                                   mlir::None);
+    } else {
+      return block_args.front();
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h]: `block_args` is a really confusing name here. I know this is used in the entire file, probably for historic reasons. `args` would be much better.

File: tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h [hunk 6]
```diff
@@ -148,44 +155,173 @@ inline Optional<CmpIPredicate> getIntCmpPredicate(
...
+  return nullptr;
+}
+
+template <>
+inline Value* MapLhloOpToStdScalarOp<xla_lhlo::SignOp>(
+    xla_lhlo::SignOp lhlo_op, ArrayRef<Type> result_types,
+    ArrayRef<Value*> block_args, OpBuilder b) {
+  Type element_type = block_args.front()->getType();
+  Value* one = b.create<mlir::ConstantFloatOp>(
+      lhlo_op.getLoc(), APFloat(1.0), FloatType::getF32(b.getContext()));
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h]: You can use `b.getFloatAttribute(element_type, 1.0)` here. The construction of one should be moved into the conditional.

File: tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h [hunk 7]
```diff
@@ -148,44 +155,173 @@ inline Optional<CmpIPredicate> getIntCmpPredicate(
...
+  if (element_type.isa<FloatType>()) {
+    Operation* op = b.create<::mlir::CopySignOp>(lhlo_op.getLoc(), result_types,
+                                                 one, block_args[0]);
+    return op->getResult(0);
+  }
+  return nullptr;
+}
+
+template <>
+inline Value* MapLhloOpToStdScalarOp<xla_lhlo::TanhOp>(
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h]: Would it make sense to have a template for all the operations that map directly to a float operation?

**Review by sherhut — CHANGES_REQUESTED** [review]
Thanks for adding this. Please avoid using Operation * and use the specific operations type where possible.

File: tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h (hunk 7 above)
Comment (dfki-ehna) [reply tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h]: Should we do this in this PR or would it make sense to do it in another PR?

**Comment by gbaned** [comment]
@dfki-ehna Can you please resolve conflicts? Thanks!

File: tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h [hunk 8]
```diff
@@ -148,44 +163,195 @@ inline Optional<CmpIPredicate> getIntCmpPredicate(
...
+
+template <>
+inline Value MapLhloOpToStdScalarOp<xla_lhlo::SignOp>(
+    xla_lhlo::SignOp lhlo_op, ArrayRef<Type> result_types, ArrayRef<Value> args,
+    OpBuilder b) {
+  Type element_type = args.front()->getType();
+  if (element_type.isa<FloatType>()) {
+    Value one = b.create<mlir::ConstantOp>(lhlo_op.getLoc(), b.getF32Type(),
+                                           b.getFloatAttr(element_type, 1.0));
+    return b
```
Comment (pifon2a) [inline tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h]: I am curious if this return statement needs `.getResult()`. I thought that for ops with one result it is not necessary.

File: tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h [hunk 9]
```diff
@@ -148,44 +163,195 @@ inline Optional<CmpIPredicate> getIntCmpPredicate(
...
+  return nullptr;
+}
+
+template <>
+inline Value MapLhloOpToStdScalarOp<xla_lhlo::SignOp>(
+    xla_lhlo::SignOp lhlo_op, ArrayRef<Type> result_types, ArrayRef<Value> args,
+    OpBuilder b) {
+  Type element_type = args.front()->getType();
+  if (element_type.isa<FloatType>()) {
+    Value one = b.create<mlir::ConstantOp>(lhlo_op.getLoc(), b.getF32Type(),
```
Comment (pifon2a) [inline tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h]: nit: introduce `loc = lhlo_op.getLoc();`

File: tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h [hunk 10]
```diff
@@ -59,69 +64,79 @@ template <typename LHLO_BinaryOp>
 using ScalarIOp = typename ScalarOp<LHLO_BinaryOp>::IOp;
 
 template <typename LhloOp>
-Operation* MapLhloOpToStdScalarOp(LhloOp lhlo_op, ArrayRef<Type> result_types,
-                                  ArrayRef<Value> block_args, OpBuilder b) {
-  Type element_type = block_args.front().getType();
+Value MapLhloOpToStdScalarOp(LhloOp lhlo_op, ArrayRef<Type> result_types,
```
Comment (pifon2a) [inline tensorflow/compiler/mlir/xla/transforms/map_lhlo_to_scalar_op.h]: Thanks for changing the result type. It became cleaner.

File: tensorflow/compiler/mlir/xla/transforms/lhlo_legalize_to_linalg.cc [hunk 11]
```diff
@@ -17,20 +17,20 @@ limitations under the License.
...
-#include "mlir/IR/Builders.h"  // TF:llvm-project
-#include "mlir/IR/Function.h"  // TF:llvm-project
-#include "mlir/IR/Location.h"  // TF:llvm-project
-#include "mlir/IR/MLIRContext.h"  // TF:llvm-project
-#include "mlir/IR/Operation.h"  // TF:llvm-project
-#include "mlir/IR/PatternMatch.h"  // TF:llvm-project
-#include "mlir/IR/StandardTypes.h"  // TF:llvm-project
-#include "mlir/Pass/Pass.h"  // TF:llvm-project
-#include "mlir/Transforms/DialectConversion.h"  // TF:llvm-project
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"    // TF:local_config_mlir
```
Comment (pifon2a) [inline tensorflow/compiler/mlir/xla/transforms/lhlo_legalize_to_linalg.cc]: I am pretty sure these lines should not be changed.

**Review by pifon2a — DISMISSED** [review]
Approving, assuming that ConvertOp will support all conversions from/to index/float/int.

**Comment by gbaned** [comment]
@dfki-ehna  Can you please resolve conflicts? Thanks!

**Comment by pifon2a** [comment]
@gbaned The problem is one the MLIR side. We are working on it.

**Comment by gbaned** [comment]
@pifon2a Any update on this PR, please. Thanks!

**Comment by rthadur** [comment]
Seems auto-merge is not happening but the changes are now committed so we can close this. Thank you for the PR.

**Pull request closed without merging**